### ISSUE TITLE
Bug with wcs.WCS.rotateCD?

### ISSUE THREAD

**Issue by gbrammer** [astropy/astropy#5175]
The rotateCD method appears to assume a transposed version of the WCS object's own CD matrix attribute.  A proposed fix would be the following, flipping the signs of the sines:

```python
def rotateCD(self, theta):
    _theta = np.deg2rad(theta)
    _mrot = np.zeros(shape=(2, 2), dtype=np.double)
    _mrot[0] = (np.cos(_theta), -np.sin(_theta))
    _mrot[1] = (np.sin(_theta), np.cos(_theta))
    new_cd = np.dot(self.wcs.cd, _mrot)
    self.wcs.cd = new_cd
```

Here is some test code that demonstrates the problem, where the rotated pixel scale of rectangular pixels isn't preserved.
```python
import astropy.wcs as pywcs

## Nominal rectangular WFC3/IR pixel
cd_wfc3 = np.array([[  2.35945978e-05,   2.62448998e-05],
                    [  2.93050803e-05,  -2.09858771e-05]])

## Square pixel
cd_square = np.array([[0.1/3600., 0], [0, 0.1/3600.]])

for cd, label in zip([cd_wfc3, cd_square], ['WFC3/IR', 'Square']):
    wcs = pywcs.WCS()
    wcs.wcs.cd = cd
    wcs.rotateCD(45.)
    print '%s pixel: pre=%s, rot=%s' %(label,
                        np.sqrt((cd**2).sum(axis=0))*3600, 
                        np.sqrt((wcs.wcs.cd**2).sum(axis=0))*3600)
# WFC3/IR pixel: pre=[ 0.1354  0.121 ], rot=[ 0.1282  0.1286]
# Square  pixel: pre=[ 0.1  0.1], rot=[ 0.1  0.1]
```

**Comment by pllim** [comment]
c/c @sosey since WFC3/IR is mentioned.

**Comment by mcara** [comment]
I am not discussing whether or not there is a bug in the ``rotateCD`` but the example provided and the associated solution are not very clear... 

Specifically, the solution suggests that switching the sign of the ``sin(theta)`` terms would fix the supposed bug. If that is the case, your fix should be equivalent to using ``rotateCD(-theta)``. I just want to ask you if running existing code with ``-theta`` fixes the issue.

**Comment by gbrammer** [comment]
You're right of course.  That's just the convention on the direction of rotation, which is fine the way it is.  What I neglected to mention before was that the order of arguments to the dot product appears to be backwards.  In `wcs.py`, it's 
```python
new_cd = np.dot(self.wcs.cd, _mrot)
```
where it should be 
```python
new_cd = np.dot(_mrot, self.wcs.cd)
```

**Comment by mcara** [comment]
Sorry for the delay. I just checked that np.dot(A, B) is indeed matrix multiplication between A*B and not transpose(A)*B as @gbrammer claimed ("assume a transposed version of the WCS object's own CD matrix"). In my opinion, the question is: 1) do we rotate the image coordinate system (CS) first and then apply CD matrix, OR 2) do we rotate the (generally skewed) intermediate CS. Currently ``rotateCD`` seems to be doing the first approach and what @gbrammer is proposing (as well as @nden in #5189) is to perform the rotation in the intermediate pixel coordinates. For the definition of of "intermediate pixel coordinates" please refer to Fig. 1 in http://www.atnf.csiro.au/people/mcalabre/WCS/wcs.pdf 

Personally I do not use ``rotateCD`` at all and I do not think there is STScI code relying on it. So, if people agree that rotations should be performed in the intermediate plane - then go for it.

Also, due to the ambiguity in the definition of the "rotated CD" (in which order rotations should be performed?) and also because, in my opinion, ``astropy.wcs`` should not be providing mechanisms/tools for "transforming" WCS, I would propose to remove this function from ``astropy.wcs``.

However, if anyone is relying on ``rotateCD()``, in order to preserve compatibility with any existing code, I would propose that ``rotateCD()`` should have a parameter such as ``intermediate_pixel_plane=False`` which would return old result and if ``intermediate_pixel_plane=True``, ``rotateCD`` would return ``np.dot(_mrot, self.wcs.cd)``.

**Comment by mcara** [comment]
Upon further investigation, it appears this function (``rotateCD``) was designed to be used by the ``recenter()`` function once available in ``PyWCS`` package (see https://www.stsci.edu/svn/ssb/stsci_python/stsdas/tags/stsdas_3_4/stsdas/pkg/analysis/dither/opydriz/wcsutil.py assuming the link works for outside users). I would assume that that old code was not buggy and that ``rotateCD`` was doing what original implementers wanted it to do. This again points to the problem of defining how rotations are being applied (=in what order) and ``recenter()`` show that there are uses for the current definition. For this reason, I would prefer the approach of introducing a parameter such as ``intermediate_pixel_plane=False`` (or a shorter version) as described in my previous comment if this function is to be kept alive.

**Comment by nden** [comment]
@mcara I am confused by your last statement. `pywcs` was not used in the old `pydrizzle`. In fact the WCS was not used in the old `pydrizzle`. Let's focus on the current software. 

After thinking about this a bit more, :+1: on deprecating this method as it is confusing and inaccurate. The FITS standard uses the PC matrix (the CD matrix being one of the supported conventions). The problem with the CD matrix is that it's not possible to separate scaling from rotation. 

Looking at the case of applying the CD/PC matrix and then rotation:
In the presence of distortions the intermediate frame is after the PC matrix and before CDELT is applied. The CD matrix transform goes directly to physical coordinates and I'm not sure what the meaning of the rotation is there. I could see how a PC matrix can be rotated but not a CD.

The case of applying rotation and then CD/PC matrix looks wrong to me in both cases because of the shift by CRPIX. Rotation could be applied before that shift but then it's part of the distortion. However, I don't think it should be applied between the shift with CRPIX and the CD/PC matrix.

Are there any objections to deprecating and eventually removing this method?

**Comment by mcara** [comment]
@nden ``pydrizzle`` has nothing to do with what I said in my previous comment. What I said was that  ``rotateCD()`` was introduced in ``PyWCS`` as an auxiliary function called from within ``PyWCS.recenter()``. That was the only place where I could find this function being used at all. Most likely it was OK for *that specific purpose* but because of the ambiguities that we keep mentioning, I think it is difficult to make this function generally applicable for most use cases.

As a side note I do not see any problem (or much difference) between "rotating" CD or PC matrices. What I do not like is the notion of rotating a matrix. I think the thinking should be about rotating the image (or coordinate system) and what effect we want to achieve.

In any case, I all for deprecating this function.

**Comment by bsipocz** [comment]
This may be closed now given that #5240 is merged?
